onnxruntime/contrib_ops/webgpu/bert/flash_attention.cc

```diff
@@ -0,0 +1,424 @@
+      // the waves in the workgroup, instead of leaving them idle.
+      for (var q_idx = 0u; q_idx < TILE_SIZE && q_idx_start + q_idx < uniforms.new_sequence_length; q_idx++)
+      {
+          // Leveraging the subgroups for parallelism, compute dot product of QK.
+          // We validate q_idx,wave_id to be less than TILE_SIZE, computeDotProduct only needs to
+          // validate sg_id as being less than QKV_HEAD_VECTORIZED_SIZE.
+          computeDotProduct(q_idx, wave_id, sg_id, sg_size);
+      }
+    }
+    // Insert barrier before SoftMax reads the dot product values across K.
+    workgroupBarrier();
+
+    let wave_lane_valid:bool = q_idx_global_using_wave_valid && sg_id < TILE_SIZE && sg_id + k_start < uniforms.present_sequence_length;
+    computeSoftMax(wave_id, sg_id, wave_lane_valid);
+    computeO(wave_id, sg_id, wave_lane_valid);
+}
+workgroupBarrier();
+if (q_idx_global_using_wave_valid)
+{
+  writeo(wave_id, q_idx_global, head_idx, sg_id, sg_size);
+}
+)MAIN_FN";
+
+  return Status::OK();
+}
+
+Status ApplyFlashAttention(const Tensor* Q, const Tensor* K, const Tensor* V, const Tensor* attention_bias,
+                           Tensor* output, const Tensor* past_key, Tensor* present_key, const Tensor* past_value, Tensor* present_value,
+                           const WebgpuAttentionParameters& parameters, onnxruntime::webgpu::ComputeContext& context) {
+  ORT_RETURN_IF_ERROR(CopyKVCache(context, parameters, K, past_key, present_key, V, past_value, present_value, parameters.past_sequence_length_, parameters.total_sequence_length_));
+
+  const uint32_t subgroup_size = context.MinSubgroupSize();
+  const uint32_t tile_size = subgroup_size;
+  bool has_attention_bias = attention_bias != nullptr;
+  FlashAttentionProgram program{"FlashAttention", has_attention_bias, subgroup_size, tile_size, parameters.head_size_, parameters.num_heads_};
+  program.AddInputs({{Q, ProgramTensorMetadataDependency::TypeAndRank, 4},
+                     {present_key, ProgramTensorMetadataDependency::TypeAndRank, 4},
+                     {present_value, ProgramTensorMetadataDependency::TypeAndRank, 4},
+                     {attention_bias, ProgramTensorMetadataDependency::TypeAndRank}});
+  program.AddOutputs({{output, ProgramTensorMetadataDependency::TypeAndRank, 4}});
+  const float alpha = parameters.scale_ == 0.0f ? 1.f / sqrt(static_cast<float>(parameters.head_size_))
+                                                : parameters.scale_;
+  std::string cache_hint = std::to_string(has_attention_bias) +
+                           std::to_string(subgroup_size) +
+                           std::to_string(tile_size) +
+                           std::to_string(parameters.head_size_) +
+                           std::to_string(parameters.num_heads_);
+  program.SetDispatchGroupSize(parameters.num_heads_, (parameters.sequence_length_ + tile_size - 1) / tile_size, 1)
```

Comment:
Hi，@sushraja-msft，I have two questions about apply the algorithm in Phi3 model, 
1. From generating the second tokens, `sequence_length_` is always 1, so the DispatchGroupSize is always 32 (`num_heads_`), Because the kernels in the same block must be executed in the same EU(intel gpu) | SM (NV gpu), there are only 32 EUs | SMs used in this algorithm. If the driver arranges some blocks in one EU | SM, or `num_heads_` is smaller in other models, there are fewer EUs | SMs used. Does it mean that the algorithm performance is the same running on a 32 EUs | SMs gpu and a 128 EUs | SMs gpu?
2. The algorithm is mainly a loop to tile the KVs and compute the results. When generating the 256th token, the count of loop is 8(256/32(tile_size)). The count of loop is 16 for the 512th token, and 32 for the 1024th token. Does it means that generating the 1024th token costs 4 times time than generating the 256th token?

I also drafted a FlashAttention-V2 in JS EP, https://github.com/microsoft/onnxruntime/pull/22915, and want to apply it to SD model. I think FlashAttention[V1|V2] algorithm is used when `sequence_length_` and `kv_sequence_length_` are all big size. Because the `Prompt token` is 501 in the test, so we could see a good prefill speed. If only `kv_sequence_length_` is large size, `FlashAttention decoding` may be better, some details is here https://crfm.stanford.edu/2023/10/12/flashdecoding.html

Reply:
xhcao - Thank you for contributing a FA2 implementation in WebGPU. I'll take a look at your implemenation this weekend but getting FA to work in webgpu without subgroup support is a remarkable feat. 

I think the point you are making in 1/2 are the same in that dispatch is not parallelizing beyond batch/head size. Yes, this is true and is a limitation in Flash Attention 1. This specific issue was addressed in FA 2- https://arxiv.org/pdf/[2307.08691](https://arxiv.org/pdf/2307.08691), Paragraph 3.2.

My goal with this change was to land changes incrementally, first land a basic implementation for MHA, then GQA then upgrade to FA2.

Finally, though I did observe the speed up earlier with this change - this week after a rebase baseline MHA got faster - I now see FA being just 1tk/s faster than regular MHA for a sequence length of 501 tokens. This prompted further investigation.

Per operator performance numbers show that during prefill Attention is just 8% of the time spend - at 22 tk's each token takes 45ms on an Intel Alderlake device. Even if I have 0 cost attention 8% of 45ms will save 4ms and result in just 2tk/s speed up. Net, I am not sure what the authors meant by " This yields 2-4× wall-clock time speedup over optimized baselines, up to 10-20× memory saving" in https://arxiv.org/pdf/[2307.08691](https://arxiv.org/pdf/2307.08691). Is it 2-4x total inference time speedup or attention wall clock time speedup. I am investigating to see if combined with static KV cache this change will show more performance wins before I pursue this change further.

Do let me know what perf improvements you are seeing with FA2 for stable diffusion.

Reply:
In FA2 Paper, the author said  in Paragraph,
 ` We also parallelize
over the batch dimension and number of heads dimension, as done in FlashAttention. The increased
parallelism over sequence length helps improve occupancy (fraction of GPU resources being used) when the
batch size and number of heads are small, leading to speedup in this case.`
But for phi3 model, in decoding stage, the sequence length is 1, so the parallelizing issue is still not fixed. And in prefill stage, in your test, sequence length is 501, so there is no parallelizing issue here.
But if the Prompt tokens are 20,  parallelizing issue also exists, if you test this situation, you may get an unexpected result.
you referred the conclusion in the paper " This yields 2-4× wall-clock time speedup over optimized baselines, up to 10-20× memory saving", I think it was related with FA1, comparing with standard attention implementing by naive algorithm.
Not like language models, the sequence length in SD is always very large. Currently, upstream code cannot run SD 2.1 with attention operators, chrome will be out of memory.